File: src/test/java/com/mateusz/rest/webservices/restful_web_services/jpa/UserJpaResourceTest.java
package com.mateusz.rest.webservices.restful_web_services.jpa;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.http.MediaType;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

import com.fasterxml.jackson.databind.ObjectMapper;

import com.mateusz.rest.webservices.restful_web_services.jpa.PostRepository;
import com.mateusz.rest.webservices.restful_web_services.jpa.UserJpaResource;
import com.mateusz.rest.webservices.restful_web_services.jpa.UserRepository;
import com.mateusz.rest.webservices.restful_web_services.security.SpringSecurityConfiguration;
import com.mateusz.rest.webservices.restful_web_services.users.Post;
import com.mateusz.rest.webservices.restful_web_services.users.User;

import static org.hamcrest.Matchers.containsString;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@WebMvcTest(UserJpaResource.class)
@Import(SpringSecurityConfiguration.class) // to avoid 403 forbidden error
@WithMockUser
public class UserJpaResourceTest {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private UserRepository userRepository;

    @MockBean
    private PostRepository postRepository;

    @Autowired
    private ObjectMapper objectMapper; // to handle LocalDate once and for all

    @Test
    public void testRetrieveAllUsers() throws Exception {
        List<User> mockUsers = List.of(
                new User(1, "Test User 1", LocalDate.of(1990, 1, 1)),
                new User(2, "Test User 2", LocalDate.of(1992, 2, 2)));
        given(userRepository.findAll()).willReturn(mockUsers);

        mockMvc.perform(get("/users"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$[0].user_name").value("Test User 1"))
                .andExpect(jsonPath("$[1].user_name").value("Test User 2"));
    }

    @Test
    public void testRetrieveOneUser() throws Exception {
        User mockUser = new User(1, "Test User", LocalDate.of(1990, 1, 1));
        given(userRepository.findById(1)).willReturn(Optional.of(mockUser));

        mockMvc.perform(get("/users/1"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.user_name").value("Test User"));
    }

    @Test
    public void testCreateOneUser() throws Exception {
        User newUser = new User(null, "New User", LocalDate.of(1995, 5, 5));
        User savedUser = new User(3, "New User", LocalDate.of(1995, 5, 5));
        given(userRepository.save(any(User.class))).willReturn(savedUser);

        mockMvc.perform(post("/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(newUser)))
                .andExpect(status().isCreated())
                .andExpect(header().string("Location", containsString("/users/3")));
    }

    @Test
    public void testDeleteOneUser() throws Exception {
        mockMvc.perform(MockMvcRequestBuilders.delete("/users/{id}", 1))
                .andExpect(status().isOk());

        // verify that the repository's deleteById method was called
        verify(userRepository).deleteById(1);
    }

    @Test
    public void testRetrieveAllPostsForUser_UserExists() throws Exception {
        User mockUser = mock(User.class);
        Post post1 = new Post();
        post1.setDescription("I love my job :)");
        Post post2 = new Post();
        post2.setDescription("Another post");
        List<Post> mockPosts = List.of(post1, post2);

        // stubbing methods on the mock objects
        given(userRepository.findById(1)).willReturn(Optional.of(mockUser));
        given(mockUser.getPosts()).willReturn(mockPosts);

        mockMvc.perform(get("/users/{id}/posts", 1))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$[0].description").value("I love my job :)"))
                .andExpect(jsonPath("$[1].description").value("Another post"));
    }

    @Test
    public void testRetrieveAllPostsForUser_UserDoesNotExist() throws Exception {
        given(userRepository.findById(anyInt())).willReturn(Optional.empty());

        mockMvc.perform(get("/users/{id}/posts", 1))
                .andExpect(status().isNotFound());
    }

    @Test
    public void testCreateOnePostForUser_UserExists() throws Exception {
        User mockUser = new User(1, "Test User", LocalDate.of(1990, 1, 1));
        Post newPost = new Post();
        newPost.setDescription("New Post Description");
        Post savedPost = new Post();
        savedPost.setId(10); // assuming saved post gets ID of 10
        savedPost.setDescription(newPost.getDescription());
        savedPost.setUser(mockUser);

        // stubbing
        given(userRepository.findById(1)).willReturn(Optional.of(mockUser));
        given(postRepository.save(any(Post.class))).willReturn(savedPost);

        mockMvc.perform(post("/users/{id}/posts", 1)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(newPost)))
                .andExpect(status().isCreated())
                .andExpect(header().string("Location", containsString("/users/1/posts/10")));
    }

    @Test
    public void testCreateOnePostForUser_UserDoesNotExist() throws Exception {
        // stub userRepository to return "empty Optional" indicating "user not found"
        given(userRepository.findById(anyInt())).willReturn(Optional.empty());

        Post newPost = new Post();
        newPost.setDescription("New Post Description");

        mockMvc.perform(post("/users/{id}/posts", 1)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(newPost)))
                .andExpect(status().isNotFound());
    }

    @Test
    public void testRetrieveOnePost_PostExists() throws Exception {
        User mockUser = new User(1, "Test User", LocalDate.of(1990, 1, 1));
        Post mockPost = new Post(1, "Test Post Description");
        mockPost.setUser(mockUser);

        // Stubbing
        given(userRepository.findById(1)).willReturn(Optional.of(mockUser));
        given(postRepository.findById(1)).willReturn(Optional.of(mockPost));

        mockMvc.perform(get("/users/{userId}/posts/{postId}", 1, 1))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.description").value("Test Post Description"))
                .andExpect(jsonPath("$._links.user.href", containsString("/users/1")))
                .andExpect(jsonPath("$._links.all-posts.href", containsString("/users/1/posts")));
    }

    @Test
    public void testRetrieveOnePost_PostDoesNotExist() throws Exception {
        // stub userRepository to return a valid user
        User mockUser = new User(1, "Test User", LocalDate.of(1990, 1, 1));
        given(userRepository.findById(1)).willReturn(Optional.of(mockUser));

        // stub postRepository to simulate post not found
        given(postRepository.findById(anyInt())).willReturn(Optional.empty());

        mockMvc.perform(get("/users/{userId}/posts/{postId}", 1, 99)) // Assuming 99 is a non-existent post ID
                .andExpect(status().isNotFound());
    }
} // ./mvnw -Dtest=UserJpaResourceTest test

File: src/test/kotlin/com/mateusz/rest/webservices/restful_web_services/spring_boot_social_media_restful_api/jpa/UserJpaResourceTest.kt
package com.mateusz.rest.webservices.restful_web_services.spring_boot_social_media_restful_api.jpa

import com.fasterxml.jackson.databind.ObjectMapper
import org.hamcrest.Matchers
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers
import org.mockito.BDDMockito
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.context.annotation.Import
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import java.time.LocalDate
import java.util.*

import java.util.Optional

import org.springframework.security.test.context.support.WithMockUser
import com.mateusz.rest.webservices.restful_web_services.spring_boot_social_media_restful_api.security.SpringSecurityConfiguration
import com.mateusz.rest.webservices.restful_web_services.spring_boot_social_media_restful_api.users.Post
import com.mateusz.rest.webservices.restful_web_services.spring_boot_social_media_restful_api.users.User

@WebMvcTest(UserJpaResource::class)
@Import(SpringSecurityConfiguration::class) // to avoid 403 forbidden error
@WithMockUser
@Suppress("UNCHECKED_CAST")
class UserJpaResourceTest {
    @Autowired
    private lateinit var mockMvc: MockMvc

    @MockBean
    private lateinit var userRepository: UserRepository

    @MockBean
    private lateinit var postRepository: PostRepository

    @Autowired
    private lateinit var objectMapper: ObjectMapper


    @Test
    @Throws(Exception::class)
    fun testRetrieveAllUsers() {
        val mockUsers: List<User> = listOf(
            User(1, "Test User 1", LocalDate.of(1990, 1, 1)),
            User(2, "Test User 2", LocalDate.of(1992, 2, 2))
        )
        BDDMockito.given(userRepository.findAll()).willReturn(mockUsers)

        mockMvc.perform(MockMvcRequestBuilders.get("/users"))
            .andExpect(MockMvcResultMatchers.status().isOk())
            .andExpect(MockMvcResultMatchers.jsonPath("$[0].user_name").value("Test User 1"))
            .andExpect(MockMvcResultMatchers.jsonPath("$[1].user_name").value("Test User 2"))
    }
    

    @Test
    @Throws(Exception::class)
    fun testRetrieveOneUser() {
        val mockUser = User(1, "Test User", LocalDate.of(1990, 1, 1))
        BDDMockito.given(userRepository.findById(1)).willReturn(Optional.of(mockUser) as Optional<User?>)

        mockMvc.perform(MockMvcRequestBuilders.get("/users/1"))
            .andExpect(MockMvcResultMatchers.status().isOk())
            .andExpect(MockMvcResultMatchers.jsonPath("$.user_name").value("Test User"))
    }

    @Test
    @Throws(Exception::class)
    fun testCreateOneUser() {
        val newUser = User(null, "New User", LocalDate.of(1995, 5, 5))
        val savedUser = User(3, "New User", LocalDate.of(1995, 5, 5))
        BDDMockito.given(userRepository.save(ArgumentMatchers.any(User::class.java))).willReturn(savedUser)

        mockMvc.perform(
            MockMvcRequestBuilders.post("/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(newUser))
        )
            .andExpect(MockMvcResultMatchers.status().isCreated())
            .andExpect(MockMvcResultMatchers.header().string("Location", Matchers.containsString("/users/3")))
    }

    @Test
    @Throws(Exception::class)
    fun testDeleteOneUser() {
        mockMvc.perform(MockMvcRequestBuilders.delete("/users/{id}", 1))
            .andExpect(MockMvcResultMatchers.status().isOk())

        // verify that the repository's deleteById method was called
        Mockito.verify(userRepository)?.deleteById(1)
    }

    @Test
    @Throws(Exception::class)
    fun testRetrieveAllPostsForUser_UserExists() {
        val mockUser: User = Mockito.mock(User::class.java)
        val post1 = Post()
        post1.description = "I love my job :)"
        val post2 = Post()
        post2.description = "Another post"
        val mockPosts: List<Post> = listOf(post1, post2)

        // stubbing methods on the mock objects
        BDDMockito.given(userRepository.findById(1)).willReturn(Optional.of(mockUser) as Optional<User?>)
        BDDMockito.given(mockUser.posts).willReturn(mockPosts)

        mockMvc.perform(MockMvcRequestBuilders.get("/users/{id}/posts", 1))
            .andExpect(MockMvcResultMatchers.status().isOk())
            .andExpect(MockMvcResultMatchers.jsonPath("$[0].description").value("I love my job :)"))
            .andExpect(MockMvcResultMatchers.jsonPath("$[1].description").value("Another post"))
    }

    @Test
    @Throws(Exception::class)
    fun testRetrieveAllPostsForUser_UserDoesNotExist() {
        BDDMockito.given(userRepository.findById(ArgumentMatchers.anyInt())).willReturn(Optional.empty<User>() as Optional<User?>)

        mockMvc.perform(MockMvcRequestBuilders.get("/users/{id}/posts", 1))
            .andExpect(MockMvcResultMatchers.status().isNotFound())
    }

    @Test
    @Throws(Exception::class)
    fun testCreateOnePostForUser_UserExists() {
        val mockUser = User(1, "Test User", LocalDate.of(1990, 1, 1))
        val newPost = Post()
        newPost.description = "New Post Description"
        val savedPost = Post()
        savedPost.id = 10
        savedPost.description = newPost.description
        savedPost.user = mockUser

        // stubbing
        BDDMockito.given(userRepository.findById(1)).willReturn(Optional.of(mockUser) as Optional<User?>)
        BDDMockito.given(postRepository.save(ArgumentMatchers.any(Post::class.java))).willReturn(savedPost)

        mockMvc.perform(
            MockMvcRequestBuilders.post("/users/{id}/posts", 1)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(newPost))
        )
            .andExpect(MockMvcResultMatchers.status().isCreated())
            .andExpect(MockMvcResultMatchers.header().string("Location", Matchers.containsString("/users/1/posts/10")))
    }

    @Test
    @Throws(Exception::class)
    fun testCreateOnePostForUser_UserDoesNotExist() {
        // stub userRepository to return "empty Optional" indicating "user not found"
        val emptyOptional: Optional<User> = Optional.empty()
        BDDMockito.given(userRepository.findById(ArgumentMatchers.anyInt())).willReturn(emptyOptional as Optional<User?>)

        val newPost = Post()
        newPost.description = "New Post Description"

        mockMvc.perform(
            MockMvcRequestBuilders.post("/users/{id}/posts", 1)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(newPost))
        )
            .andExpect(MockMvcResultMatchers.status().isNotFound())
    }

    @Test
    @Throws(Exception::class)
    fun testRetrieveOnePost_PostExists() {
        val mockUser = User(1, "Test User", LocalDate.of(1990, 1, 1))
        val mockPost = Post(1, "Test Post Description")
        mockPost.user = mockUser

        // Stubbing
        BDDMockito.given(userRepository.findById(1)).willReturn(Optional.of(mockUser) as Optional<User?>)
        BDDMockito.given(postRepository.findById(1)).willReturn(Optional.of(mockPost) as Optional<Post?>)

        mockMvc.perform(MockMvcRequestBuilders.get("/users/{userId}/posts/{postId}", 1, 1))
            .andExpect(MockMvcResultMatchers.status().isOk())
            .andExpect(MockMvcResultMatchers.jsonPath("$.description").value("Test Post Description"))
            .andExpect(MockMvcResultMatchers.jsonPath("$._links.user.href", Matchers.containsString("/users/1")))
            .andExpect(
                MockMvcResultMatchers.jsonPath(
                    "$._links.all-posts.href",
                    Matchers.containsString("/users/1/posts")
                )
            )
    }

    @Test
    @Throws(Exception::class)
    fun testRetrieveOnePost_PostDoesNotExist() {
        // stub userRepository to return a valid user
        val mockUser = User(1, "Test User", LocalDate.of(1990, 1, 1))
        

        BDDMockito.given(userRepository.findById(1)).willReturn(Optional.of(mockUser) as Optional<User?>)

        // stub postRepository to simulate post not found
        BDDMockito.given(postRepository.findById(ArgumentMatchers.anyInt())).willReturn(Optional.empty<Post>() as Optional<Post?>)

        mockMvc.perform(
            MockMvcRequestBuilders.get(
                "/users/{userId}/posts/{postId}",
                1,
                99
            )
        ) // assuming 99 is a non-existent post ID
            .andExpect(MockMvcResultMatchers.status().isNotFound())
    }
}